### DISCUSSION TITLE
Converting 32 bit stream to 16 bit

### DISCUSSION THREAD

**Discussion by wegscheider** [pschatzmann/arduino-audio-tools#629] [Q&A] [answered]
Hi Phil:

First of all, thank you very much for the fantastic library. In a project where I want to save the output from an I2S microphone to SD card, I am running into the following problem: When I artificially generate the output (via SineWaveGenerator) in 16 bit everything works perfect. However, the microphone produces a 32 bit stream which I obviously need to convert to 16 bit. That's where I am struggling. I used the following code to test but receive only zeros on the serial monitor:

#include "AudioTools.h"

uint16_t sample_rate=44100;
                                          
SineWaveGenerator<int32_t> sine_wave(32000);                         // subclass of SoundGenerator with max amplitude of 32000
GeneratedSoundStream<int32_t> in_stream(sine_wave);                  // Stream generated from sine wave
CsvStream<int16_t> out(Serial, 1);                                  // Output to Serial
NumberFormatConverterStreamT<int32_t, int16_t> conv(in_stream);
StreamCopy copier(out, conv);

void setup(){
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);  

  sine_wave.begin(1, sample_rate, N_B4);
  conv.begin();
  in_stream.begin();

  out.begin();
}

void loop(){
  copier.copy();
}

What am I doing wrong with the converter function?

Any help is very much appreciated.

Werner

**Comment by pschatzmann** [comment]
Did you try to output the 32 bit GeneratedSoundStream directly. I propose that you leave the amplitude away in the constructor of the sine_wave or increase it to the max value of a 32 bit number.
First I just want to make sure that your sound generation is working...

**Reply by wegscheider** [reply]
Yes, the GeneratedSoundStream works just fine.

**Comment by pschatzmann** [comment, marked as answer]
I was running your sketch with my proposed change with
SineWaveGenerator<int32_t> sine_wave;
And for me it works like a charm:
![image](https://user-images.githubusercontent.com/2078916/219852970-4a2714c8-6fd0-493b-9014-328bfba20d97.png)

When you use 32000 as max amplitude the converted signal is oscillating around +-0.00001490116 which gives 0 when converted to an integer!

**Reply by wegscheider** [reply]
Thank you very much! I can reproduce your result. That explains also why I got only so small numbers when using the I2S microphone. 

Now for the I2S microphone which gives me 24 bit values on a 32 bit stream, what do I need to use the full 16 bit range and not only have values from 0 to 255 in my 16 bit stream? I tried to use <int24_t> streams in my sine_wave example (amplitude 8000000) but the values are still only covering 8 bits.

**Comment by pschatzmann** [comment]
Quite recently I added some functionality to determine the max number by data type. Initially I relied on the size, but this does not work for int24_t because this has 4 bytes like the int32_t. 
I just committed the related change to the NumberFormatConverterStreamT which should work now with int24_t  in a better way.

If the volume is still too low you can add a VolumeStream to your chain....

**Comment by wegscheider** [comment]
Thank you very much, then I will try that.

**Reply by pschatzmann** [reply]
I was just converting your example to 24bit and I am getting the expected result

![image](https://user-images.githubusercontent.com/2078916/219856078-ebedb411-9cad-4cae-b83b-03e6540da8fc.png)

**Reply by wegscheider** [reply]
Great, got everything to work now. However the 24bit input from the microphone didn't work but 32 bit and a VolumeStream, then conversion to 16bit did the trick!
